Design Draft for `gh repo edit`

g14a opened the pull request:
This design draft and a dummy flow exposes more repository settings as discussed in https://github.com/cli/cli/issues/995

Attaching a google document for the design. And the head branch contains a dummy flow for `gh repo edit`

Might close #995 

Design doc - [Google doc](https://docs.google.com/document/d/1ot3GTPh7s44_7Fx26UjShfPvFhncWcf4lpCla4vmwbg/edit?usp=sharing)

Attaching screenshots for reference:

<img width="689" alt="image" src="https://user-images.githubusercontent.com/17702388/123090721-1b8da900-d446-11eb-939b-6ef8a21edb10.png">

<img width="710" alt="image" src="https://user-images.githubusercontent.com/17702388/123090910-555eaf80-d446-11eb-82ef-62fd8400237b.png">

<img width="710" alt="image" src="https://user-images.githubusercontent.com/17702388/123091639-3ad90600-d447-11eb-8be0-c2aac069c971.png">

g14a committed:
draft pr for gh repo edit

g14a committed:
draft flow for gh repo edit

g14a committed:
Merge branch 'trunk' of https://github.com/cli/cli into draft-flow-gh-repo-edit

g14a committed:
add delete head branch option

g14a commented:
@mislav | @vilmibm As a response to this comment https://github.com/cli/cli/issues/995#issuecomment-864251069, I've made a design document(link in the header description), a reference to the APIs supported by Github and a mock up dummy flow in the head branch itself.

Would love some feedback on this :)

vilmibm reviewed (changes requested):
Awesome work on this delivery! I love how thorough the google doc is.

I chatted with our designer and we agree that given how many things we might expose here, it might be best to do a multiselect to see what the user wants to change instead of asking about each setting.

Mind another design pass that mirrors what pr/issue edit do?

g14a commented:
@vilmibm thanks for the feedback :) I think this would be a mixture of multi select and a boolean survey.

For eg: There are boolean settings like `Enable Wiki ?` or `Convert to a template repository ?`. In this case it is redundant to ask the user to first select if he wants to edit these changes and again ask him if his answer is a yes or a no. We can directly ask him in a confirm prompt whether or not he's going to go for it.

I propose we keep the text fields and options which can have multiple choices like repo visibility as multi select i.e 

- Name
- Description
- Homepage URL
- Visibility
- Merge Options
 
 and boolean settings as a confirm prompt i.e 
 
- Enable Issues
- Enable Wiki
- Converting to a template repo
- Archive Repo
- Delete head branches after merging PR

Would like to hear out your thoughts as well on this.

g14a commented:
any thoughts @vilmibm ?

g14a commented:
Hey @vilmibm just wanted to do a follow up about your thoughts on this :)

vilmibm commented:
Hey @g14a ,

Thanks for your patience, sorry for the long delay.

Good point about the bool prompts. I like the compromise you suggested--feel free to go for it!

g14a commented:
Thanks for the reply @vilmibm I updated the google doc with the heading `2nd iteration of Design Draft`. I figured out there would be a scenario where the user would want to directly update only the boolean fields so I added another category called `Other Options` under which there are boolean prompts. In this way he doesn't need to go through the other prompts in order to reach the boolean prompts.

Google doc - [doc](https://docs.google.com/document/d/1ot3GTPh7s44_7Fx26UjShfPvFhncWcf4lpCla4vmwbg/edit?usp=sharing)

Attaching screenshot for reference:

<img width="1081" alt="image" src="https://user-images.githubusercontent.com/17702388/127098013-c1007776-a2cb-4e11-91fd-dda394e5f721.png">

Let me know if any changes are need to be made. If the design is convincing, I can start working on the implementation :)

g14a commented:
Hey @vilmibm just wanted to get a follow up on the final design iteration :)

g14a commented:
Hey @vilmibm I don't mean to hurry but did you find a moment to go through the latest draft attached above? Thank you :)

mislav commented:
Hi @g14a, sorry for no communication! We were busy with making other features for gh and we didn't prioritize the `edit` command yet.

The prompt-based `repo edit` that you've sketched out looks good; it seems that it follows the existing design of `gh issue edit`.

I would suggest, though, to start with a non-interactive `repo edit` implementation that just takes flags for each field to edit. Once the non-interactive `edit` is done, you could make an interactive one in a subsequent PR. I think this would simplify the implementation & review process by splitting out a potentially large undertaking into two discrete bodies of work. Thank you!

g14a commented:
@mislav thank you! That's all I needed. On it.

g14a commented:
The actual PR can be found at https://github.com/cli/cli/pull/4318 as this is a design draft.

mislav commented:
@g14a Now that the non-interactive portion has been merged https://github.com/cli/cli/pull/4318, you're welcome to explore the interactive parts again if you find time. However, if you will create a new PR for that, then this one can be closed. Thank you! 🙇
